libs/gltfio/src/StbProvider.cpp

```diff
@@ -245,23 +271,123 @@ const char* StbProvider::getPopMessage() const {
+    const auto rowSrcIncrement = srcRowsPerDstRow * sourceLineStride;
+    const auto pixelSrcIncrement = srcColsPerDstCol * componentsPerPixel;
+    auto destination = std::unique_ptr<uint8_t[]>(new uint8_t[destinationLineStride * destinationHeight]);
+    for (auto destinationY = 0u ; destinationY < destinationHeight ; ++destinationY) {
+        const auto rowSrcBase = destinationY * rowSrcIncrement;
+        auto pixelSource = rowSrcBase;
+        const auto rowDstBase = destinationY * destinationLineStride;
+        const auto rowDstEnd = rowDstBase + rowRelativeDstEnd;
+        for (auto pixelDst = rowDstBase; pixelDst < rowDstEnd; pixelDst += componentsPerPixel) {
+            for (auto component = 0u ; component < componentsPerPixel ; ++component) {
+                auto sum = sumBaseline;
+                const auto componentSource = pixelSource + component;
+                for (auto srcCompOffset = 0u ; srcCompOffset < srcPixelsPerDstPixel ; ++srcCompOffset) {
+                    sum += (uint32_t)source[componentSource + srcCompOffsets[srcCompOffset]];
+                }
+                destination[pixelDst + component] = sum / srcPixelsPerDstPixel;
+            }
+            pixelSource += pixelSrcIncrement;
+        }
+    }
+
+    return destination;
+}
+
+/**
+ * Scales down a single byte per component RGBA image where each destination pixel corresponds to
+ * a whole number of source pixels on each dimension. canSimpleScaleDown(sourceWidth, sourceHeight, destinationWidth, destinationHeight)
+ * must return true as a precondition.
+ */
+static std::unique_ptr<uint8_t[]> simpleScaleDownRgba(
+        const uint8_t *source, uint32_t sourceWidth, uint32_t sourceHeight, uint32_t sourceLineStride,
+        uint32_t destinationWidth, uint32_t destinationHeight, uint32_t destinationLineStride) {
+    return simpleScaleDown<4u>(source, sourceWidth, sourceHeight, sourceLineStride,
+                               destinationWidth, destinationHeight, destinationLineStride);
+}
+
+void StbProvider::decode(TextureInfo& info) {
+    auto& source = info.sourceBuffer;
+    int width, height, comp;
+
+    std::uint8_t* texels = static_cast<std::uint8_t*>(stbi_load_from_memory(
+            source.data(), source.size(), &width, &height, &comp, 4));
+    source.clear();
+    source.shrink_to_fit();
+    if (texels == nullptr) {
+        info.decodedTexelsBaseMipmap.store(DECODING_ERROR);
+        return;
+    }
+    info.resized = ((width != info.texture->getWidth()) || (height != info.texture->getHeight()));
+    if (info.resized) {
+        if (canSimpleScaleDown(width, height,
+                info.texture->getWidth(), info.texture->getHeight())) {
+            auto scaled_down_texels = simpleScaleDownRgba(
+                    texels, width, height, width * sizeof(std::uint8_t) * 4u,
+                    info.texture->getWidth(), info.texture->getHeight(),
+                    info.texture->getWidth() * sizeof(std::uint8_t) * 4u);
```

Comment:
why are you using SimpleScaleDownRgba instead of always using image::resampleImage?

Is it for performance? Does it matter? What are the odds that given a max size, an image can be resized by a integer amount?

Also, when you take the canSimpleScalDown path, you're not first transforming to linear, so your averaging is wrong. At any rate, both algorithm wouldn't produce the same result and I think they should.

Reply:
Hi pixelflinger. Yes, this is for performance reasons which we are pretty sensitive to in our application. Typically this will be used to take our power of 2 dimension textures and scale them down by a factor of 2 or 4 (for example 2048 to 1024 or maybe 512).
So this basically averages 4 source texels or 16 source texels to determine the value of the destination pixel. You are correct that this computation is done in sRGB and not linear scale so the answer will not be 100% precise, however I don't think the difference would be discernible to most users.
A conversion to linear, running `SimpleScaleDownRgba()` followed by a conversion back to sRGB is doable, but would be considerably more computationally expensive. Alternatively, if you are pretty confident the performance difference between this and `image::resampleImage()` would be negligible even in the dimension special case needed for `SimpleScaleDownRgba()`, we could simply use `image::resampleImage()` in all cases and remove `SimpleScaleDownRgba()`.

Reply:
> A conversion to linear, running SimpleScaleDownRgba() followed by a conversion back to sRGB is doable, but would be considerably more computationally expensive.

It is, but it's also the correct thing to do for color data. Filament is properly color managed everywhere, and we need to preserve that.

Reply:
I just did some profiling w/o the simple scale down method (using resample) and the results for scaling down an 8k example texture down to 4k are:
- ios: simple scale down ~100ms, rescale >2m
- android: simple scale down ~200ms, rescale >5m
- wasm: simple scale down ~250ms, rescale >5m

admittedly the textures we're working with are always a power of 2, so simple scale down works on the fly, but using rescale is prohibitively expensive. note that all of those numbers were using release builds, I can provide the exact texture if you'd like to test it yourself.

even just conversions from/to sRBG/linar take considerably long time
```
toLinearWithAlpha took 747 ms
fromLinearTosRGB took 164 ms
```

do those numbers look correct to you or are we doing something wrong here?

Reply:
Given those numbers we really need to fix `image::resampleImage`, because if someone uses this new API and ends up in the slow path, they'll think the program is hung. That just seems unacceptable.

Given this situation, I think it would be better for the resizing to be done by the app, so that it's not our (filament's) responsibility how the scaling is done.

I think, instead of specifying a max texture size, we could pass a texture transform callback (use utils::Invocable), if the callback is provided, gltfio will call it for all textures. The callback can decide if it wants to resize the texture and do so if needed.

I realize this would change this PR significantly, but I think it's a better solution. @romainguy what do you think?

Reply:
that sounds reasonable to me, that gives us enough flexibility, we can keep this downsampling logic on our side. will make those changes tomorrow